Add json support to istioctl validate

pawelbeza opened the pull request:
**Please provide a description of this PR:**
Fixes: #46135

Previously istioctl validate didn't support directories consisting of json files due to the following guard `if !info.IsDir() && filepath.Ext(path) == ".yaml"`. After this patch `istioctl validate` will support json files similarly as `istioctl analyze`

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [X] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [X] Test and Release
- [X] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

istio-testing commented:
Hi @pawelbeza. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

ymesika commented on istioctl/pkg/validate/validate.go:
```diff
@@ -300,9 +308,17 @@ func validateFiles(istioNamespace *string, defaultNamespace string, filenames []
+			if isFileFormatValid(path) {
 				processFile(path)
+			} else {
+				log.Warnf("skipping file %v, recognized file extensions are: %v", path, fileExtensions)
```
This is a bit of a changed behavior because I'm not sure we should warn about irrelevant files (like before)

hanxiaop replied:
I think we don't need this warning, as we only validate resources that we care about. If we warn about every unrecognized file, we may end up with many spam logs when validating the directory.

pawelbeza replied:
so should I just skip them silently?

pawelbeza replied:
the reason why I marked it as warn is that there is similar part of the code in `istioctl analyze` which throws message to stderr for irrelevant files https://github.com/istio/istio/blob/dfcf08db3fd306335771b5fe1b1f3038d85dec3f/istioctl/pkg/analyze/analyze.go#L400

hanxiaop replied:
I think it's fine to remove the warning since `analyze` provides more exhaustive info, even for like parsing errors, and here we are just validating Istio configs.

hanxiaop reviewed (changes requested):
We need to add some tests for this

pawelbeza commented:
@ymesika @hanxiaop I've removed unnecessary warn and added tests. Tests might look scary but they're analogous to the the already existing YAML tests.

hanxiaop reviewed (commented):
Could you please also add a releasenote since this contains user-facing changes?

hanxiaop commented:
/ok-to-test

istio-testing commented:
@pawelbeza: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 25412c45a5f77c410044a553783615a941dd2b06 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/46136/release-notes_istio/1683667454032613376) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hanxiaop commented:
> @ymesika @hanxiaop I've removed unnecessary warn and added tests.
> 
> JSON tests might look scary but they're pretty much the same as existing YAML tests.

Tests look good, thanks. We may revise the configs to testdata files instead of strings as a follow-up. But LGTM if adding a releasenote.

ymesika commented:
LGTM but as it's a UX we will need the release note.

hanxiaop reviewed (approved):
Nice!

hanxiaop commented on releasenotes/notes/46136.yaml:
```diff
@@ -0,0 +1,12 @@
+
+# issue is a list of GitHub issues resolved in this note.
+issue:
+  - https://github.com/istio/istio/issues/46136
```
nit: this can just be either `https://github.com/istio/istio/issues/46136` or `46136`
